### 5.5) Gathering Input Graphs

In [1]:
!pip install torch-scatter -f https://data.pyg.org/whl/torch-2.5.1+cu124.html
!pip install torch-sparse -f https://data.pyg.org/whl/torch-2.5.1+cu124.html
!pip install torch-geometric

Looking in links: https://data.pyg.org/whl/torch-2.5.1+cu124.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 88.8 MB/s eta 0:00:00:00:01:01
Looking in links: https://data.pyg.org/whl/torch-2.5.1+cu124.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 41.0 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.1/63.1 kB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 20.0 MB/s eta 0:00:00a 0:00:01


In [2]:
import torch
from torch_geometric.data import Data
from torch.serialization import add_safe_globals
import os

# Allow PyTorch to unpickle torch_geometric.data.Data objects
add_safe_globals([Data])

# Then load the files
graph_dir = "/kaggle/input/chicken-productivity-rate-graphs/graph_data_zip"
pt_files = [f for f in os.listdir(graph_dir) if f.endswith('.pt')]

graphs = [
    torch.load(os.path.join(graph_dir, f), weights_only=False)
    for f in pt_files
]

/usr/local/lib/python3.11/dist-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /usr/local/lib/python3.11/dist-packages/torch_scatter/_scatter_cuda.so: undefined symbol: _ZN2at4_ops16div__Tensor_mode4callERNS_6TensorERKS2_St8optionalIN3c1017basic_string_viewIcEEE
  warnings.warn(f"An issue occurred while importing 'torch-scatter'. "
/usr/local/lib/python3.11/dist-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /usr/local/lib/python3.11/dist-packages/torch_sparse/_spmm_cuda.so: undefined symbol: _ZN5torch8autograd12VariableInfoC1ERKN2at6TensorE
  warnings.warn(f"An issue occurred while importing 'torch-sparse'. "


In [3]:
graphs

[Data(x=[560, 560], edge_index=[2, 440258], edge_attr=[440258, 4], y=[1]),
 Data(x=[603, 603], edge_index=[2, 319768], edge_attr=[319768, 4], y=[1]),
 Data(x=[525, 525], edge_index=[2, 92854], edge_attr=[92854, 4], y=[1]),
 Data(x=[648, 648], edge_index=[2, 477458], edge_attr=[477458, 4], y=[1]),
 Data(x=[484, 484], edge_index=[2, 327346], edge_attr=[327346, 4], y=[1]),
 Data(x=[430, 430], edge_index=[2, 295144], edge_attr=[295144, 4], y=[1]),
 Data(x=[716, 716], edge_index=[2, 348140], edge_attr=[348140, 4], y=[1]),
 Data(x=[706, 706], edge_index=[2, 401906], edge_attr=[401906, 4], y=[1]),
 Data(x=[577, 577], edge_index=[2, 488282], edge_attr=[488282, 4], y=[1]),
 Data(x=[599, 599], edge_index=[2, 582338], edge_attr=[582338, 4], y=[1]),
 Data(x=[451, 451], edge_index=[2, 274074], edge_attr=[274074, 4], y=[1]),
 Data(x=[426, 426], edge_index=[2, 277242], edge_attr=[277242, 4], y=[1]),
 Data(x=[591, 591], edge_index=[2, 571268], edge_attr=[571268, 4], y=[1]),
 Data(x=[199, 199], edge_in

In [4]:
# padding so model can expect a fixed set of dimensions from graphs

max_node_dim = max(g.x.shape[1] for g in graphs)
max_edge_dim = max(g.edge_attr.shape[1] for g in graphs)

def pad_features(graph, max_node_dim, max_edge_dim):
    # Pad node features
    node_feat = graph.x
    if node_feat.shape[1] < max_node_dim:
        pad_size = max_node_dim - node_feat.shape[1]
        padding = torch.zeros((node_feat.shape[0], pad_size), dtype=node_feat.dtype)
        graph.x = torch.cat([node_feat, padding], dim=1)
    
    # Pad edge features
    edge_feat = graph.edge_attr
    if edge_feat.shape[1] < max_edge_dim:
        pad_size = max_edge_dim - edge_feat.shape[1]
        padding = torch.zeros((edge_feat.shape[0], pad_size), dtype=edge_feat.dtype)
        graph.edge_attr = torch.cat([edge_feat, padding], dim=1)
    
    return graph

In [5]:
graphs_padded = [pad_features(g, max_node_dim, max_edge_dim) for g in graphs]

### 6) Create GAT Model for Graph Regression¶

GAT model takes graph as input, applies GAT layers to learn node embeddings, aggregates (mean pooling or attention pooling) to a graph level embedding, and passes that to fully connected layers to predict a single value (productivity)

global mean pooling

In [50]:
import torch
import torch.nn.functional as F
from torch_geometric.nn import GATConv, global_mean_pool
from torch_geometric.loader import DataLoader
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

class GATGraphRegressor(torch.nn.Module):
    def __init__(self, in_node_feats, hidden_dim=64, heads=4):
        super(GATGraphRegressor, self).__init__()
        self.gat1 = GATConv(in_node_feats, hidden_dim, heads=heads, concat=True, dropout=0.2)
        self.gat2 = GATConv(hidden_dim * heads, hidden_dim, heads=1, concat=True, dropout=0.2)

        self.mlp = torch.nn.Sequential(
            torch.nn.Linear(hidden_dim, hidden_dim),
            torch.nn.ReLU(),
            torch.nn.Linear(hidden_dim, 1)
        )

    def forward(self, data):
        x, edge_index, batch = data.x, data.edge_index, data.batch
        x = self.gat1(x, edge_index)
        x = F.elu(x)
        x = self.gat2(x, edge_index)
        x = F.elu(x)
        x = global_mean_pool(x, batch)
        return self.mlp(x).squeeze(1)

def train_epoch(model, loader, criterion, optimizer, device):
    model.train()
    losses = []
    for batch in loader:
        batch = batch.to(device)
        optimizer.zero_grad()
        out = model(batch)
        loss = criterion(out, batch.y.view(-1))
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()
        losses.append(loss.item())
    return np.mean(losses)

def eval_model(model, loader, criterion, device):
    model.eval()
    losses = []
    preds, targets = [], []
    with torch.no_grad():
        for batch in loader:
            batch = batch.to(device)
            out = model(batch)
            loss = criterion(out, batch.y.view(-1))
            losses.append(loss.item())
            preds.extend(out.cpu().numpy())
            targets.extend(batch.y.view(-1).cpu().numpy())
    rmse = mean_squared_error(targets, preds, squared=False)
    mae = mean_absolute_error(targets, preds)
    return np.mean(losses), rmse, mae

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
kf = KFold(n_splits=5, shuffle=True, random_state=0)

all_rmse = []
all_mae = []

for fold, (train_idx, val_idx) in enumerate(kf.split(graphs_padded)):
    print(f"Fold {fold + 1}")
    train_dataset = [graphs_padded[i] for i in train_idx]
    val_dataset = [graphs_padded[i] for i in val_idx]

    train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)

    model = GATGraphRegressor(in_node_feats=max_node_dim).to(device)
    criterion = torch.nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)

    best_val_loss = float('inf')
    patience = 10
    patience_counter = 0

    for epoch in range(100):
        train_loss = train_epoch(model, train_loader, criterion, optimizer, device)
        val_loss, rmse, mae = eval_model(model, val_loader, criterion, device)
        print(f"Epoch {epoch} | Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f} | RMSE: {rmse:.4f} | MAE: {mae:.4f}")

        if val_loss < best_val_loss:
            best_val_loss = val_loss
            patience_counter = 0
            torch.save(model.state_dict(), f"best_model_fold{fold}.pt")
        else:
            patience_counter += 1
            if patience_counter >= patience:
                print("Early stopping")
                break

    # Load best model and evaluate on validation set
    model.load_state_dict(torch.load(f"best_model_fold{fold}.pt"))
    _, rmse, mae = eval_model(model, val_loader, criterion, device)
    all_rmse.append(rmse)
    all_mae.append(mae)
    print(f"Fold {fold + 1} final RMSE: {rmse:.4f}, MAE: {mae:.4f}")

print(f"\nAverage RMSE: {np.mean(all_rmse):.4f} ± {np.std(all_rmse):.4f}")
print(f"Average MAE: {np.mean(all_mae):.4f} ± {np.std(all_mae):.4f}")


Fold 1
Epoch 0 | Train Loss: 1.0966 | Val Loss: 0.7271 | RMSE: 0.8527 | MAE: 0.7572
Epoch 1 | Train Loss: 1.0875 | Val Loss: 0.7656 | RMSE: 0.8750 | MAE: 0.7719
Epoch 2 | Train Loss: 1.0909 | Val Loss: 0.8673 | RMSE: 0.9313 | MAE: 0.8030
Epoch 3 | Train Loss: 1.2166 | Val Loss: 0.9918 | RMSE: 0.9959 | MAE: 0.8392
Epoch 4 | Train Loss: 1.1289 | Val Loss: 0.9736 | RMSE: 0.9867 | MAE: 0.8255
Epoch 5 | Train Loss: 1.1602 | Val Loss: 0.9578 | RMSE: 0.9787 | MAE: 0.8086
Epoch 6 | Train Loss: 1.1021 | Val Loss: 0.8801 | RMSE: 0.9382 | MAE: 0.7701
Epoch 7 | Train Loss: 1.0085 | Val Loss: 0.7621 | RMSE: 0.8730 | MAE: 0.7102
Epoch 8 | Train Loss: 0.9108 | Val Loss: 0.5686 | RMSE: 0.7541 | MAE: 0.6253
Epoch 9 | Train Loss: 0.8200 | Val Loss: 0.4713 | RMSE: 0.6865 | MAE: 0.5622
Epoch 10 | Train Loss: 0.7599 | Val Loss: 0.3812 | RMSE: 0.6174 | MAE: 0.5101
Epoch 11 | Train Loss: 0.6646 | Val Loss: 0.2838 | RMSE: 0.5328 | MAE: 0.4719
Epoch 12 | Train Loss: 0.5779 | Val Loss: 0.2086 | RMSE: 0.4568 | M

global_max_pool instead of global_mean_pool

In [51]:
import torch
import torch.nn.functional as F
from torch_geometric.nn import GATConv, global_max_pool
from torch_geometric.loader import DataLoader
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

class GATGraphRegressor(torch.nn.Module):
    def __init__(self, in_node_feats, hidden_dim=64, heads=4):
        super(GATGraphRegressor, self).__init__()
        self.gat1 = GATConv(in_node_feats, hidden_dim, heads=heads, concat=True, dropout=0.2)
        self.gat2 = GATConv(hidden_dim * heads, hidden_dim, heads=1, concat=True, dropout=0.2)

        self.mlp = torch.nn.Sequential(
            torch.nn.Linear(hidden_dim, hidden_dim),
            torch.nn.ReLU(),
            torch.nn.Linear(hidden_dim, 1)
        )

    def forward(self, data):
        x, edge_index, batch = data.x, data.edge_index, data.batch
        x = self.gat1(x, edge_index)
        x = F.elu(x)
        x = self.gat2(x, edge_index)
        x = F.elu(x)
        x = global_max_pool(x, batch)
        return self.mlp(x).squeeze(1)

# Normalize targets
ys_all = [g.y.item() for g in graphs_padded]
mean_y = np.mean(ys_all)
std_y = np.std(ys_all)

for g in graphs_padded:
    g.y = torch.tensor([(g.y.item() - mean_y) / std_y], dtype=torch.float)

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

kf = KFold(n_splits=5, shuffle=True, random_state=0)

def train_epoch(model, loader, criterion, optimizer):
    model.train()
    losses = []
    for batch in loader:
        batch = batch.to(device)
        optimizer.zero_grad()
        out = model(batch)
        loss = criterion(out, batch.y.view(-1))
        loss.backward()
        optimizer.step()
        losses.append(loss.item())
    return np.mean(losses)

def eval_model(model, loader, criterion):
    model.eval()
    losses = []
    preds, targets = [], []
    with torch.no_grad():
        for batch in loader:
            batch = batch.to(device)
            out = model(batch)
            loss = criterion(out, batch.y.view(-1))
            losses.append(loss.item())
            preds.extend(out.cpu().numpy())
            targets.extend(batch.y.view(-1).cpu().numpy())
    rmse = mean_squared_error(targets, preds, squared=False)
    mae = mean_absolute_error(targets, preds)
    return np.mean(losses), rmse, mae

all_rmse = []
all_mae = []

for fold, (train_idx, val_idx) in enumerate(kf.split(graphs_padded)):
    print(f"Fold {fold + 1}")
    train_dataset = [graphs_padded[i] for i in train_idx]
    val_dataset = [graphs_padded[i] for i in val_idx]

    train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)

    model = GATGraphRegressor(in_node_feats=max_node_dim).to(device)
    criterion = torch.nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)

    best_val_loss = float('inf')
    patience = 10
    patience_counter = 0

    for epoch in range(100):
        train_loss = train_epoch(model, train_loader, criterion, optimizer)
        val_loss, rmse, mae = eval_model(model, val_loader, criterion)
        print(f"Epoch {epoch} | Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f} | RMSE: {rmse:.4f} | MAE: {mae:.4f}")

        if val_loss < best_val_loss:
            best_val_loss = val_loss
            patience_counter = 0
            torch.save(model.state_dict(), f"best_model_max_fold{fold}.pt")
        else:
            patience_counter += 1
            if patience_counter >= patience:
                print("Early stopping triggered.")
                break

    model.load_state_dict(torch.load(f"best_model_max_fold{fold}.pt"))
    _, rmse, mae = eval_model(model, val_loader, criterion)
    all_rmse.append(rmse)
    all_mae.append(mae)
    print(f"Fold {fold + 1} final RMSE: {rmse:.4f}, MAE: {mae:.4f}")

print(f"\nAverage RMSE: {np.mean(all_rmse):.4f} ± {np.std(all_rmse):.4f}")
print(f"Average MAE: {np.mean(all_mae):.4f} ± {np.std(all_mae):.4f}")

Fold 1


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 0 | Train Loss: 1.0928 | Val Loss: 0.6418 | RMSE: 0.8011 | MAE: 0.7292


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 1 | Train Loss: 1.0855 | Val Loss: 0.6413 | RMSE: 0.8008 | MAE: 0.7277


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 2 | Train Loss: 1.0776 | Val Loss: 0.6461 | RMSE: 0.8038 | MAE: 0.7265


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 3 | Train Loss: 1.0657 | Val Loss: 0.6352 | RMSE: 0.7970 | MAE: 0.7207


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 4 | Train Loss: 1.0455 | Val Loss: 0.6323 | RMSE: 0.7952 | MAE: 0.7152


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 5 | Train Loss: 1.0168 | Val Loss: 0.5981 | RMSE: 0.7734 | MAE: 0.6951


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 6 | Train Loss: 0.9586 | Val Loss: 0.5632 | RMSE: 0.7504 | MAE: 0.6670


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 7 | Train Loss: 0.8786 | Val Loss: 0.4632 | RMSE: 0.6806 | MAE: 0.6033


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 8 | Train Loss: 0.7601 | Val Loss: 0.3583 | RMSE: 0.5985 | MAE: 0.5222


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 9 | Train Loss: 0.6570 | Val Loss: 0.2555 | RMSE: 0.5055 | MAE: 0.4226


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 10 | Train Loss: 0.5607 | Val Loss: 0.1961 | RMSE: 0.4428 | MAE: 0.3654


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 11 | Train Loss: 0.4803 | Val Loss: 0.1333 | RMSE: 0.3651 | MAE: 0.3137


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 12 | Train Loss: 0.4206 | Val Loss: 0.1220 | RMSE: 0.3492 | MAE: 0.2838


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 13 | Train Loss: 0.3951 | Val Loss: 0.1179 | RMSE: 0.3434 | MAE: 0.2800


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 14 | Train Loss: 0.3611 | Val Loss: 0.1329 | RMSE: 0.3646 | MAE: 0.2991


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 15 | Train Loss: 0.3574 | Val Loss: 0.1584 | RMSE: 0.3979 | MAE: 0.3307


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 16 | Train Loss: 0.3331 | Val Loss: 0.1280 | RMSE: 0.3577 | MAE: 0.2976


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 17 | Train Loss: 0.3414 | Val Loss: 0.1326 | RMSE: 0.3641 | MAE: 0.3119


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 18 | Train Loss: 0.3225 | Val Loss: 0.1306 | RMSE: 0.3614 | MAE: 0.3054


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 19 | Train Loss: 0.3114 | Val Loss: 0.1393 | RMSE: 0.3733 | MAE: 0.3218


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 20 | Train Loss: 0.3030 | Val Loss: 0.1453 | RMSE: 0.3812 | MAE: 0.3014


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 21 | Train Loss: 0.2939 | Val Loss: 0.1483 | RMSE: 0.3851 | MAE: 0.3346


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 22 | Train Loss: 0.3034 | Val Loss: 0.1616 | RMSE: 0.4020 | MAE: 0.3508


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 23 | Train Loss: 0.3032 | Val Loss: 0.1478 | RMSE: 0.3845 | MAE: 0.3198
Early stopping triggered.
Fold 1 final RMSE: 0.3434, MAE: 0.2800
Fold 2


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 0 | Train Loss: 1.1405 | Val Loss: 0.4438 | RMSE: 0.6662 | MAE: 0.6265


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 1 | Train Loss: 1.1347 | Val Loss: 0.4515 | RMSE: 0.6719 | MAE: 0.6333


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 2 | Train Loss: 1.1259 | Val Loss: 0.4533 | RMSE: 0.6733 | MAE: 0.6351


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 3 | Train Loss: 1.1171 | Val Loss: 0.4564 | RMSE: 0.6756 | MAE: 0.6380


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 4 | Train Loss: 1.1021 | Val Loss: 0.4536 | RMSE: 0.6735 | MAE: 0.6363


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 5 | Train Loss: 1.0713 | Val Loss: 0.4351 | RMSE: 0.6596 | MAE: 0.6223


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 6 | Train Loss: 1.0207 | Val Loss: 0.3980 | RMSE: 0.6309 | MAE: 0.5927


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 7 | Train Loss: 0.9447 | Val Loss: 0.3250 | RMSE: 0.5701 | MAE: 0.5290


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 8 | Train Loss: 0.8406 | Val Loss: 0.2369 | RMSE: 0.4867 | MAE: 0.4352


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 9 | Train Loss: 0.6843 | Val Loss: 0.1992 | RMSE: 0.4464 | MAE: 0.4050


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 10 | Train Loss: 0.5353 | Val Loss: 0.1549 | RMSE: 0.3936 | MAE: 0.3400


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 11 | Train Loss: 0.4465 | Val Loss: 0.1567 | RMSE: 0.3959 | MAE: 0.3095


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 12 | Train Loss: 0.3986 | Val Loss: 0.1738 | RMSE: 0.4169 | MAE: 0.3102


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 13 | Train Loss: 0.3627 | Val Loss: 0.1605 | RMSE: 0.4007 | MAE: 0.2906


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 14 | Train Loss: 0.3738 | Val Loss: 0.1918 | RMSE: 0.4380 | MAE: 0.3200


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 15 | Train Loss: 0.3572 | Val Loss: 0.1820 | RMSE: 0.4266 | MAE: 0.3153


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 16 | Train Loss: 0.3252 | Val Loss: 0.1943 | RMSE: 0.4407 | MAE: 0.3254


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 17 | Train Loss: 0.3127 | Val Loss: 0.1735 | RMSE: 0.4165 | MAE: 0.3010


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 18 | Train Loss: 0.3076 | Val Loss: 0.2313 | RMSE: 0.4809 | MAE: 0.3593


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 19 | Train Loss: 0.2999 | Val Loss: 0.1558 | RMSE: 0.3947 | MAE: 0.2752


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 20 | Train Loss: 0.3151 | Val Loss: 0.1730 | RMSE: 0.4160 | MAE: 0.2940
Early stopping triggered.
Fold 2 final RMSE: 0.3936, MAE: 0.3400
Fold 3


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 0 | Train Loss: 0.8249 | Val Loss: 1.7299 | RMSE: 1.3153 | MAE: 0.9405


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 1 | Train Loss: 0.8210 | Val Loss: 1.7251 | RMSE: 1.3134 | MAE: 0.9393


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 2 | Train Loss: 0.8161 | Val Loss: 1.7223 | RMSE: 1.3124 | MAE: 0.9370


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 3 | Train Loss: 0.8097 | Val Loss: 1.7116 | RMSE: 1.3083 | MAE: 0.9340


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 4 | Train Loss: 0.7981 | Val Loss: 1.6954 | RMSE: 1.3021 | MAE: 0.9274


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 5 | Train Loss: 0.7819 | Val Loss: 1.6655 | RMSE: 1.2906 | MAE: 0.9161


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 6 | Train Loss: 0.7472 | Val Loss: 1.5992 | RMSE: 1.2646 | MAE: 0.8957


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 7 | Train Loss: 0.6854 | Val Loss: 1.5112 | RMSE: 1.2293 | MAE: 0.8547


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 8 | Train Loss: 0.5851 | Val Loss: 1.3793 | RMSE: 1.1744 | MAE: 0.7849


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 9 | Train Loss: 0.4695 | Val Loss: 1.2282 | RMSE: 1.1082 | MAE: 0.6635


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 10 | Train Loss: 0.3617 | Val Loss: 1.0276 | RMSE: 1.0137 | MAE: 0.5746


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 11 | Train Loss: 0.2858 | Val Loss: 0.9311 | RMSE: 0.9650 | MAE: 0.5071


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 12 | Train Loss: 0.2401 | Val Loss: 0.8614 | RMSE: 0.9281 | MAE: 0.4828


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 13 | Train Loss: 0.2195 | Val Loss: 0.7916 | RMSE: 0.8897 | MAE: 0.5038


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 14 | Train Loss: 0.2149 | Val Loss: 0.8114 | RMSE: 0.9008 | MAE: 0.4773


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 15 | Train Loss: 0.2073 | Val Loss: 0.7902 | RMSE: 0.8890 | MAE: 0.4765


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 16 | Train Loss: 0.1871 | Val Loss: 0.7875 | RMSE: 0.8874 | MAE: 0.4712


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 17 | Train Loss: 0.1853 | Val Loss: 0.7732 | RMSE: 0.8793 | MAE: 0.4655


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 18 | Train Loss: 0.1820 | Val Loss: 0.7586 | RMSE: 0.8710 | MAE: 0.4706


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 19 | Train Loss: 0.1640 | Val Loss: 0.7441 | RMSE: 0.8626 | MAE: 0.4659


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 20 | Train Loss: 0.1640 | Val Loss: 0.7705 | RMSE: 0.8778 | MAE: 0.4602


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 21 | Train Loss: 0.1657 | Val Loss: 0.7306 | RMSE: 0.8548 | MAE: 0.4620


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 22 | Train Loss: 0.1471 | Val Loss: 0.7035 | RMSE: 0.8387 | MAE: 0.4743


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 23 | Train Loss: 0.1666 | Val Loss: 0.7359 | RMSE: 0.8579 | MAE: 0.4604


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 24 | Train Loss: 0.1513 | Val Loss: 0.7105 | RMSE: 0.8429 | MAE: 0.4610


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 25 | Train Loss: 0.1351 | Val Loss: 0.7511 | RMSE: 0.8666 | MAE: 0.4587


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 26 | Train Loss: 0.1321 | Val Loss: 0.7630 | RMSE: 0.8735 | MAE: 0.4646


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 27 | Train Loss: 0.1315 | Val Loss: 0.7145 | RMSE: 0.8453 | MAE: 0.4545


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 28 | Train Loss: 0.1335 | Val Loss: 0.6997 | RMSE: 0.8365 | MAE: 0.4512


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 29 | Train Loss: 0.1354 | Val Loss: 0.7045 | RMSE: 0.8394 | MAE: 0.4510


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 30 | Train Loss: 0.1119 | Val Loss: 0.6889 | RMSE: 0.8300 | MAE: 0.4529


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 31 | Train Loss: 0.1299 | Val Loss: 0.6855 | RMSE: 0.8280 | MAE: 0.4605


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 32 | Train Loss: 0.1075 | Val Loss: 0.6800 | RMSE: 0.8246 | MAE: 0.4560


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 33 | Train Loss: 0.1087 | Val Loss: 0.6779 | RMSE: 0.8234 | MAE: 0.4614


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 34 | Train Loss: 0.1057 | Val Loss: 0.6779 | RMSE: 0.8233 | MAE: 0.4582


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 35 | Train Loss: 0.0983 | Val Loss: 0.6934 | RMSE: 0.8327 | MAE: 0.4618


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 36 | Train Loss: 0.1104 | Val Loss: 0.6965 | RMSE: 0.8346 | MAE: 0.4624


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 37 | Train Loss: 0.1086 | Val Loss: 0.7127 | RMSE: 0.8442 | MAE: 0.4684


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 38 | Train Loss: 0.0946 | Val Loss: 0.7536 | RMSE: 0.8681 | MAE: 0.4823


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 39 | Train Loss: 0.0975 | Val Loss: 0.6642 | RMSE: 0.8150 | MAE: 0.4727


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 40 | Train Loss: 0.0989 | Val Loss: 0.6784 | RMSE: 0.8237 | MAE: 0.4698


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 41 | Train Loss: 0.0887 | Val Loss: 0.6788 | RMSE: 0.8239 | MAE: 0.4718


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 42 | Train Loss: 0.0841 | Val Loss: 0.7081 | RMSE: 0.8415 | MAE: 0.4771


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 43 | Train Loss: 0.0818 | Val Loss: 0.7088 | RMSE: 0.8419 | MAE: 0.4762


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 44 | Train Loss: 0.0764 | Val Loss: 0.7026 | RMSE: 0.8382 | MAE: 0.4772


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 45 | Train Loss: 0.0747 | Val Loss: 0.7828 | RMSE: 0.8848 | MAE: 0.5057


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 46 | Train Loss: 0.0766 | Val Loss: 0.6873 | RMSE: 0.8290 | MAE: 0.4835


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 47 | Train Loss: 0.0756 | Val Loss: 0.6975 | RMSE: 0.8351 | MAE: 0.4808


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 48 | Train Loss: 0.0732 | Val Loss: 0.6787 | RMSE: 0.8238 | MAE: 0.4897


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 49 | Train Loss: 0.0631 | Val Loss: 0.7285 | RMSE: 0.8535 | MAE: 0.4886
Early stopping triggered.
Fold 3 final RMSE: 0.8150, MAE: 0.4727
Fold 4


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 0 | Train Loss: 1.1173 | Val Loss: 0.5535 | RMSE: 0.7440 | MAE: 0.6664


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 1 | Train Loss: 1.1123 | Val Loss: 0.5505 | RMSE: 0.7419 | MAE: 0.6652


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 2 | Train Loss: 1.1066 | Val Loss: 0.5464 | RMSE: 0.7392 | MAE: 0.6632


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 3 | Train Loss: 1.0971 | Val Loss: 0.5381 | RMSE: 0.7336 | MAE: 0.6590


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 4 | Train Loss: 1.0829 | Val Loss: 0.5228 | RMSE: 0.7230 | MAE: 0.6501


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 5 | Train Loss: 1.0524 | Val Loss: 0.4922 | RMSE: 0.7015 | MAE: 0.6298


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 6 | Train Loss: 1.0046 | Val Loss: 0.4396 | RMSE: 0.6630 | MAE: 0.5918


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 7 | Train Loss: 0.9115 | Val Loss: 0.3585 | RMSE: 0.5987 | MAE: 0.5219


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 8 | Train Loss: 0.7774 | Val Loss: 0.2554 | RMSE: 0.5054 | MAE: 0.4168


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 9 | Train Loss: 0.5977 | Val Loss: 0.1983 | RMSE: 0.4453 | MAE: 0.3531


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 10 | Train Loss: 0.4556 | Val Loss: 0.2162 | RMSE: 0.4650 | MAE: 0.3729


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 11 | Train Loss: 0.3817 | Val Loss: 0.2747 | RMSE: 0.5242 | MAE: 0.4408


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 12 | Train Loss: 0.3554 | Val Loss: 0.2715 | RMSE: 0.5210 | MAE: 0.4453


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 13 | Train Loss: 0.3259 | Val Loss: 0.2792 | RMSE: 0.5284 | MAE: 0.4454


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 14 | Train Loss: 0.3274 | Val Loss: 0.2949 | RMSE: 0.5431 | MAE: 0.4613


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 15 | Train Loss: 0.3243 | Val Loss: 0.2802 | RMSE: 0.5293 | MAE: 0.4417


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 16 | Train Loss: 0.3072 | Val Loss: 0.2666 | RMSE: 0.5163 | MAE: 0.4344


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 17 | Train Loss: 0.2940 | Val Loss: 0.2904 | RMSE: 0.5389 | MAE: 0.4595


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 18 | Train Loss: 0.2925 | Val Loss: 0.2644 | RMSE: 0.5142 | MAE: 0.4158


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 19 | Train Loss: 0.2901 | Val Loss: 0.2720 | RMSE: 0.5216 | MAE: 0.4367
Early stopping triggered.
Fold 4 final RMSE: 0.4453, MAE: 0.3531
Fold 5


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 0 | Train Loss: 0.8409 | Val Loss: 1.6721 | RMSE: 1.2931 | MAE: 1.0141


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 1 | Train Loss: 0.8353 | Val Loss: 1.6674 | RMSE: 1.2913 | MAE: 1.0163


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 2 | Train Loss: 0.8290 | Val Loss: 1.6603 | RMSE: 1.2885 | MAE: 1.0168


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 3 | Train Loss: 0.8208 | Val Loss: 1.6481 | RMSE: 1.2838 | MAE: 1.0151


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 4 | Train Loss: 0.8066 | Val Loss: 1.6279 | RMSE: 1.2759 | MAE: 1.0065


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 5 | Train Loss: 0.7835 | Val Loss: 1.5843 | RMSE: 1.2587 | MAE: 0.9985


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 6 | Train Loss: 0.7441 | Val Loss: 1.5060 | RMSE: 1.2272 | MAE: 0.9640


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 7 | Train Loss: 0.6730 | Val Loss: 1.3762 | RMSE: 1.1731 | MAE: 0.8924


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 8 | Train Loss: 0.5739 | Val Loss: 1.2027 | RMSE: 1.0967 | MAE: 0.7991


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 9 | Train Loss: 0.4742 | Val Loss: 1.0170 | RMSE: 1.0085 | MAE: 0.7190


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 10 | Train Loss: 0.3709 | Val Loss: 0.8604 | RMSE: 0.9276 | MAE: 0.6516


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 11 | Train Loss: 0.3145 | Val Loss: 0.7659 | RMSE: 0.8752 | MAE: 0.5875


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 12 | Train Loss: 0.2967 | Val Loss: 0.7074 | RMSE: 0.8411 | MAE: 0.5647


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 13 | Train Loss: 0.2640 | Val Loss: 0.6656 | RMSE: 0.8158 | MAE: 0.5518


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 14 | Train Loss: 0.2603 | Val Loss: 0.6613 | RMSE: 0.8132 | MAE: 0.5347


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 15 | Train Loss: 0.2494 | Val Loss: 0.6049 | RMSE: 0.7778 | MAE: 0.5474


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 16 | Train Loss: 0.2351 | Val Loss: 0.5979 | RMSE: 0.7732 | MAE: 0.5569


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 17 | Train Loss: 0.2245 | Val Loss: 0.6018 | RMSE: 0.7757 | MAE: 0.5142


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 18 | Train Loss: 0.2338 | Val Loss: 0.5613 | RMSE: 0.7492 | MAE: 0.5323


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 19 | Train Loss: 0.2153 | Val Loss: 0.5738 | RMSE: 0.7575 | MAE: 0.5129


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 20 | Train Loss: 0.2237 | Val Loss: 0.5665 | RMSE: 0.7527 | MAE: 0.5074


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 21 | Train Loss: 0.2186 | Val Loss: 0.5599 | RMSE: 0.7483 | MAE: 0.5053


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 22 | Train Loss: 0.2036 | Val Loss: 0.5669 | RMSE: 0.7529 | MAE: 0.5012


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 23 | Train Loss: 0.2185 | Val Loss: 0.5578 | RMSE: 0.7469 | MAE: 0.4980


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 24 | Train Loss: 0.2107 | Val Loss: 0.5359 | RMSE: 0.7321 | MAE: 0.5082


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 25 | Train Loss: 0.2016 | Val Loss: 0.5250 | RMSE: 0.7246 | MAE: 0.4988


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 26 | Train Loss: 0.1935 | Val Loss: 0.5348 | RMSE: 0.7313 | MAE: 0.4932


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 27 | Train Loss: 0.1868 | Val Loss: 0.5193 | RMSE: 0.7206 | MAE: 0.5158


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 28 | Train Loss: 0.1803 | Val Loss: 0.5204 | RMSE: 0.7214 | MAE: 0.5263


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 29 | Train Loss: 0.1959 | Val Loss: 0.5102 | RMSE: 0.7143 | MAE: 0.5223


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 30 | Train Loss: 0.1643 | Val Loss: 0.5164 | RMSE: 0.7186 | MAE: 0.4906


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 31 | Train Loss: 0.1698 | Val Loss: 0.5070 | RMSE: 0.7121 | MAE: 0.4982


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 32 | Train Loss: 0.1971 | Val Loss: 0.4912 | RMSE: 0.7009 | MAE: 0.4943


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 33 | Train Loss: 0.1633 | Val Loss: 0.5019 | RMSE: 0.7085 | MAE: 0.4884


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 34 | Train Loss: 0.1612 | Val Loss: 0.4893 | RMSE: 0.6995 | MAE: 0.4953


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 35 | Train Loss: 0.1604 | Val Loss: 0.4845 | RMSE: 0.6961 | MAE: 0.5075


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 36 | Train Loss: 0.1580 | Val Loss: 0.4835 | RMSE: 0.6953 | MAE: 0.5067


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 37 | Train Loss: 0.1562 | Val Loss: 0.4856 | RMSE: 0.6968 | MAE: 0.4869


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 38 | Train Loss: 0.1499 | Val Loss: 0.4813 | RMSE: 0.6938 | MAE: 0.4957


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 39 | Train Loss: 0.1512 | Val Loss: 0.4858 | RMSE: 0.6970 | MAE: 0.4805


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 40 | Train Loss: 0.1623 | Val Loss: 0.4723 | RMSE: 0.6872 | MAE: 0.4828


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 41 | Train Loss: 0.1449 | Val Loss: 0.4793 | RMSE: 0.6923 | MAE: 0.4936


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 42 | Train Loss: 0.1345 | Val Loss: 0.4855 | RMSE: 0.6968 | MAE: 0.4779


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 43 | Train Loss: 0.1398 | Val Loss: 0.4747 | RMSE: 0.6890 | MAE: 0.4863


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 44 | Train Loss: 0.1403 | Val Loss: 0.4775 | RMSE: 0.6910 | MAE: 0.4889


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 45 | Train Loss: 0.1245 | Val Loss: 0.4880 | RMSE: 0.6985 | MAE: 0.5162


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 46 | Train Loss: 0.1213 | Val Loss: 0.4698 | RMSE: 0.6854 | MAE: 0.4947


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 47 | Train Loss: 0.1223 | Val Loss: 0.4736 | RMSE: 0.6882 | MAE: 0.5074


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 48 | Train Loss: 0.1097 | Val Loss: 0.4576 | RMSE: 0.6764 | MAE: 0.4782


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 49 | Train Loss: 0.1119 | Val Loss: 0.4499 | RMSE: 0.6707 | MAE: 0.4701


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 50 | Train Loss: 0.1116 | Val Loss: 0.4528 | RMSE: 0.6729 | MAE: 0.4582


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 51 | Train Loss: 0.1035 | Val Loss: 0.4448 | RMSE: 0.6669 | MAE: 0.4713


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 52 | Train Loss: 0.1148 | Val Loss: 0.4549 | RMSE: 0.6745 | MAE: 0.4844


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 53 | Train Loss: 0.0960 | Val Loss: 0.4615 | RMSE: 0.6793 | MAE: 0.4872


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 54 | Train Loss: 0.1021 | Val Loss: 0.4413 | RMSE: 0.6643 | MAE: 0.4652


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 55 | Train Loss: 0.0926 | Val Loss: 0.4387 | RMSE: 0.6624 | MAE: 0.4824


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 56 | Train Loss: 0.0849 | Val Loss: 0.4322 | RMSE: 0.6574 | MAE: 0.4703


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 57 | Train Loss: 0.0892 | Val Loss: 0.4314 | RMSE: 0.6568 | MAE: 0.4604


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 58 | Train Loss: 0.0825 | Val Loss: 0.4356 | RMSE: 0.6600 | MAE: 0.4864


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 59 | Train Loss: 0.0836 | Val Loss: 0.4358 | RMSE: 0.6601 | MAE: 0.4821


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 60 | Train Loss: 0.0953 | Val Loss: 0.4174 | RMSE: 0.6460 | MAE: 0.4691


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 61 | Train Loss: 0.0894 | Val Loss: 0.4224 | RMSE: 0.6500 | MAE: 0.4635


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 62 | Train Loss: 0.0876 | Val Loss: 0.4129 | RMSE: 0.6426 | MAE: 0.4342


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 63 | Train Loss: 0.0866 | Val Loss: 0.4192 | RMSE: 0.6475 | MAE: 0.4437


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 64 | Train Loss: 0.0835 | Val Loss: 0.4081 | RMSE: 0.6388 | MAE: 0.4493


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 65 | Train Loss: 0.0815 | Val Loss: 0.4152 | RMSE: 0.6444 | MAE: 0.4495


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 66 | Train Loss: 0.0794 | Val Loss: 0.4164 | RMSE: 0.6453 | MAE: 0.4368


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 67 | Train Loss: 0.0625 | Val Loss: 0.4073 | RMSE: 0.6382 | MAE: 0.4561


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 68 | Train Loss: 0.0704 | Val Loss: 0.4331 | RMSE: 0.6581 | MAE: 0.4985


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 69 | Train Loss: 0.0662 | Val Loss: 0.4007 | RMSE: 0.6330 | MAE: 0.4544


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 70 | Train Loss: 0.0609 | Val Loss: 0.4218 | RMSE: 0.6495 | MAE: 0.4916


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 71 | Train Loss: 0.0620 | Val Loss: 0.4041 | RMSE: 0.6357 | MAE: 0.4395


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 72 | Train Loss: 0.0649 | Val Loss: 0.4080 | RMSE: 0.6387 | MAE: 0.4522


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 73 | Train Loss: 0.0626 | Val Loss: 0.4027 | RMSE: 0.6346 | MAE: 0.4450


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 74 | Train Loss: 0.0579 | Val Loss: 0.4203 | RMSE: 0.6483 | MAE: 0.4896


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 75 | Train Loss: 0.0601 | Val Loss: 0.4165 | RMSE: 0.6453 | MAE: 0.4869


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 76 | Train Loss: 0.0570 | Val Loss: 0.4465 | RMSE: 0.6682 | MAE: 0.5215


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 77 | Train Loss: 0.0501 | Val Loss: 0.3934 | RMSE: 0.6272 | MAE: 0.4449


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 78 | Train Loss: 0.0598 | Val Loss: 0.3914 | RMSE: 0.6257 | MAE: 0.4476


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 79 | Train Loss: 0.0537 | Val Loss: 0.4212 | RMSE: 0.6490 | MAE: 0.4867


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 80 | Train Loss: 0.0540 | Val Loss: 0.3967 | RMSE: 0.6298 | MAE: 0.4566


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 81 | Train Loss: 0.0550 | Val Loss: 0.4033 | RMSE: 0.6351 | MAE: 0.4642


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 82 | Train Loss: 0.0614 | Val Loss: 0.3922 | RMSE: 0.6263 | MAE: 0.4482


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 83 | Train Loss: 0.0458 | Val Loss: 0.3914 | RMSE: 0.6256 | MAE: 0.4262


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 84 | Train Loss: 0.0404 | Val Loss: 0.3900 | RMSE: 0.6245 | MAE: 0.4292


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 85 | Train Loss: 0.0362 | Val Loss: 0.3842 | RMSE: 0.6199 | MAE: 0.4183


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 86 | Train Loss: 0.0354 | Val Loss: 0.4562 | RMSE: 0.6754 | MAE: 0.5355


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 87 | Train Loss: 0.0541 | Val Loss: 0.3854 | RMSE: 0.6208 | MAE: 0.4624


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 88 | Train Loss: 0.0474 | Val Loss: 0.4242 | RMSE: 0.6513 | MAE: 0.5028


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 89 | Train Loss: 0.0458 | Val Loss: 0.3887 | RMSE: 0.6235 | MAE: 0.4268


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 90 | Train Loss: 0.0387 | Val Loss: 0.3764 | RMSE: 0.6135 | MAE: 0.4340


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 91 | Train Loss: 0.0445 | Val Loss: 0.3857 | RMSE: 0.6210 | MAE: 0.4449


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 92 | Train Loss: 0.0344 | Val Loss: 0.3974 | RMSE: 0.6304 | MAE: 0.4722


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 93 | Train Loss: 0.0356 | Val Loss: 0.3959 | RMSE: 0.6292 | MAE: 0.4503


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 94 | Train Loss: 0.0364 | Val Loss: 0.4221 | RMSE: 0.6497 | MAE: 0.5024


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 95 | Train Loss: 0.0344 | Val Loss: 0.4081 | RMSE: 0.6388 | MAE: 0.4769


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 96 | Train Loss: 0.0309 | Val Loss: 0.4217 | RMSE: 0.6494 | MAE: 0.4990


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 97 | Train Loss: 0.0278 | Val Loss: 0.3933 | RMSE: 0.6272 | MAE: 0.4433


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 98 | Train Loss: 0.0294 | Val Loss: 0.3954 | RMSE: 0.6288 | MAE: 0.4640


/usr/local/lib/python3.11/dist-packages/torch_geometric/warnings.py:11: UserWarning: The usage of `scatter(reduce='max')` can be accelerated via the 'torch-scatter' package, but it was not found
  warnings.warn(message)


Epoch 99 | Train Loss: 0.0377 | Val Loss: 0.4389 | RMSE: 0.6625 | MAE: 0.5200
Fold 5 final RMSE: 0.6135, MAE: 0.4340

Average RMSE: 0.5221 ± 0.1723
Average MAE: 0.3760 ± 0.0689


Use now globalAttention

In [52]:
import torch
import torch.nn.functional as F
from torch_geometric.nn import GATConv, GlobalAttention
from torch_geometric.loader import DataLoader
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

class GATGraphRegressor(torch.nn.Module):
    def __init__(self, in_node_feats, hidden_dim=64, heads=4):
        super(GATGraphRegressor, self).__init__()

        self.gat1 = GATConv(in_node_feats, hidden_dim, heads=heads, concat=True, dropout=0.2)
        self.gat2 = GATConv(hidden_dim * heads, hidden_dim, heads=1, concat=True, dropout=0.2)
        
        self.att_pool = GlobalAttention(gate_nn=torch.nn.Sequential(
            torch.nn.Linear(hidden_dim, 1),
            torch.nn.Sigmoid()
        ))
        
        self.mlp = torch.nn.Sequential(
            torch.nn.Linear(hidden_dim, hidden_dim),
            torch.nn.ReLU(),
            torch.nn.Linear(hidden_dim, 1)
        )

    def forward(self, data):
        x, edge_index, batch = data.x, data.edge_index, data.batch
        x = self.gat1(x, edge_index)
        x = F.elu(x)
        x = self.gat2(x, edge_index)
        x = F.elu(x)
        x = self.att_pool(x, batch)
        return self.mlp(x).squeeze(1)

# Normalize targets
ys_all = [g.y.item() for g in graphs_padded]
mean_y = np.mean(ys_all)
std_y = np.std(ys_all)

for g in graphs_padded:
    g.y = torch.tensor([(g.y.item() - mean_y) / std_y], dtype=torch.float)

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

kf = KFold(n_splits=5, shuffle=True, random_state=0)

def train_epoch(model, loader, criterion, optimizer):
    model.train()
    losses = []
    for batch in loader:
        batch = batch.to(device)
        optimizer.zero_grad()
        out = model(batch)
        loss = criterion(out, batch.y.view(-1))
        loss.backward()
        optimizer.step()
        losses.append(loss.item())
    return np.mean(losses)

def eval_model(model, loader, criterion):
    model.eval()
    losses = []
    preds, targets = [], []
    with torch.no_grad():
        for batch in loader:
            batch = batch.to(device)
            out = model(batch)
            loss = criterion(out, batch.y.view(-1))
            losses.append(loss.item())
            preds.extend(out.cpu().numpy())
            targets.extend(batch.y.view(-1).cpu().numpy())
    rmse = mean_squared_error(targets, preds, squared=False)
    mae = mean_absolute_error(targets, preds)
    return np.mean(losses), rmse, mae

all_rmse = []
all_mae = []

for fold, (train_idx, val_idx) in enumerate(kf.split(graphs_padded)):
    print(f"Fold {fold + 1}")
    train_dataset = [graphs_padded[i] for i in train_idx]
    val_dataset = [graphs_padded[i] for i in val_idx]

    train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)

    model = GATGraphRegressor(in_node_feats=max_node_dim).to(device)
    criterion = torch.nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)

    best_val_loss = float('inf')
    patience = 10
    patience_counter = 0

    for epoch in range(100):
        train_loss = train_epoch(model, train_loader, criterion, optimizer)
        val_loss, rmse, mae = eval_model(model, val_loader, criterion)
        print(f"Epoch {epoch} | Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f} | RMSE: {rmse:.4f} | MAE: {mae:.4f}")

        if val_loss < best_val_loss:
            best_val_loss = val_loss
            patience_counter = 0
            torch.save(model.state_dict(), f"best_model_ga_fold{fold}.pt")
        else:
            patience_counter += 1
            if patience_counter >= patience:
                print("Early stopping triggered.")
                break

    model.load_state_dict(torch.load(f"best_model_ga_fold{fold}.pt"))
    _, rmse, mae = eval_model(model, val_loader, criterion)
    all_rmse.append(rmse)
    all_mae.append(mae)
    print(f"Fold {fold + 1} final RMSE: {rmse:.4f}, MAE: {mae:.4f}")

print(f"\nAverage RMSE: {np.mean(all_rmse):.4f} ± {np.std(all_rmse):.4f}")
print(f"Average MAE: {np.mean(all_mae):.4f} ± {np.std(all_mae):.4f}")

Fold 1


/usr/local/lib/python3.11/dist-packages/torch_geometric/deprecation.py:26: UserWarning: 'nn.glob.GlobalAttention' is deprecated, use 'nn.aggr.AttentionalAggregation' instead
  warnings.warn(out)


Epoch 0 | Train Loss: 1.0948 | Val Loss: 0.6307 | RMSE: 0.7942 | MAE: 0.7261
Epoch 1 | Train Loss: 1.0821 | Val Loss: 0.6284 | RMSE: 0.7927 | MAE: 0.7219
Epoch 2 | Train Loss: 1.0476 | Val Loss: 0.6133 | RMSE: 0.7831 | MAE: 0.7084
Epoch 3 | Train Loss: 0.9792 | Val Loss: 0.5307 | RMSE: 0.7285 | MAE: 0.6632
Epoch 4 | Train Loss: 0.8630 | Val Loss: 0.4420 | RMSE: 0.6648 | MAE: 0.5995
Epoch 5 | Train Loss: 0.7123 | Val Loss: 0.2898 | RMSE: 0.5383 | MAE: 0.4973
Epoch 6 | Train Loss: 0.5579 | Val Loss: 0.1870 | RMSE: 0.4324 | MAE: 0.3685
Epoch 7 | Train Loss: 0.4529 | Val Loss: 0.1568 | RMSE: 0.3960 | MAE: 0.3044
Epoch 8 | Train Loss: 0.4025 | Val Loss: 0.1599 | RMSE: 0.3999 | MAE: 0.2898
Epoch 9 | Train Loss: 0.3796 | Val Loss: 0.1585 | RMSE: 0.3981 | MAE: 0.2843
Epoch 10 | Train Loss: 0.3622 | Val Loss: 0.1745 | RMSE: 0.4177 | MAE: 0.2946
Epoch 11 | Train Loss: 0.3565 | Val Loss: 0.2215 | RMSE: 0.4706 | MAE: 0.3531
Epoch 12 | Train Loss: 0.3535 | Val Loss: 0.2002 | RMSE: 0.4475 | MAE: 0.3

/usr/local/lib/python3.11/dist-packages/torch_geometric/deprecation.py:26: UserWarning: 'nn.glob.GlobalAttention' is deprecated, use 'nn.aggr.AttentionalAggregation' instead
  warnings.warn(out)


Epoch 0 | Train Loss: 1.1484 | Val Loss: 0.4295 | RMSE: 0.6554 | MAE: 0.6155
Epoch 1 | Train Loss: 1.1327 | Val Loss: 0.4385 | RMSE: 0.6622 | MAE: 0.6235
Epoch 2 | Train Loss: 1.1017 | Val Loss: 0.4347 | RMSE: 0.6593 | MAE: 0.6220
Epoch 3 | Train Loss: 1.0455 | Val Loss: 0.3961 | RMSE: 0.6293 | MAE: 0.5919
Epoch 4 | Train Loss: 0.9565 | Val Loss: 0.3074 | RMSE: 0.5544 | MAE: 0.5152
Epoch 5 | Train Loss: 0.8137 | Val Loss: 0.1977 | RMSE: 0.4446 | MAE: 0.3965
Epoch 6 | Train Loss: 0.6491 | Val Loss: 0.1041 | RMSE: 0.3227 | MAE: 0.2675
Epoch 7 | Train Loss: 0.5038 | Val Loss: 0.0804 | RMSE: 0.2836 | MAE: 0.2203
Epoch 8 | Train Loss: 0.4274 | Val Loss: 0.0794 | RMSE: 0.2818 | MAE: 0.1983
Epoch 9 | Train Loss: 0.4015 | Val Loss: 0.1186 | RMSE: 0.3443 | MAE: 0.2306
Epoch 10 | Train Loss: 0.3935 | Val Loss: 0.0960 | RMSE: 0.3098 | MAE: 0.2257
Epoch 11 | Train Loss: 0.3717 | Val Loss: 0.1539 | RMSE: 0.3923 | MAE: 0.2670
Epoch 12 | Train Loss: 0.3785 | Val Loss: 0.1299 | RMSE: 0.3604 | MAE: 0.2

/usr/local/lib/python3.11/dist-packages/torch_geometric/deprecation.py:26: UserWarning: 'nn.glob.GlobalAttention' is deprecated, use 'nn.aggr.AttentionalAggregation' instead
  warnings.warn(out)


Epoch 0 | Train Loss: 0.8351 | Val Loss: 1.6825 | RMSE: 1.2971 | MAE: 0.9423
Epoch 1 | Train Loss: 0.8196 | Val Loss: 1.6763 | RMSE: 1.2947 | MAE: 0.9312
Epoch 2 | Train Loss: 0.7915 | Val Loss: 1.6388 | RMSE: 1.2801 | MAE: 0.9105
Epoch 3 | Train Loss: 0.7375 | Val Loss: 1.5341 | RMSE: 1.2386 | MAE: 0.8647
Epoch 4 | Train Loss: 0.6371 | Val Loss: 1.3809 | RMSE: 1.1751 | MAE: 0.7827
Epoch 5 | Train Loss: 0.5038 | Val Loss: 1.1896 | RMSE: 1.0907 | MAE: 0.6544
Epoch 6 | Train Loss: 0.3781 | Val Loss: 1.0053 | RMSE: 1.0027 | MAE: 0.5590
Epoch 7 | Train Loss: 0.2997 | Val Loss: 0.8800 | RMSE: 0.9381 | MAE: 0.4863
Epoch 8 | Train Loss: 0.2634 | Val Loss: 0.8137 | RMSE: 0.9020 | MAE: 0.4637
Epoch 9 | Train Loss: 0.2486 | Val Loss: 0.7879 | RMSE: 0.8876 | MAE: 0.4520
Epoch 10 | Train Loss: 0.2363 | Val Loss: 0.7575 | RMSE: 0.8703 | MAE: 0.4503
Epoch 11 | Train Loss: 0.2339 | Val Loss: 0.7476 | RMSE: 0.8646 | MAE: 0.4531
Epoch 12 | Train Loss: 0.2281 | Val Loss: 0.7469 | RMSE: 0.8642 | MAE: 0.4

/usr/local/lib/python3.11/dist-packages/torch_geometric/deprecation.py:26: UserWarning: 'nn.glob.GlobalAttention' is deprecated, use 'nn.aggr.AttentionalAggregation' instead
  warnings.warn(out)


Epoch 0 | Train Loss: 1.1113 | Val Loss: 0.5601 | RMSE: 0.7484 | MAE: 0.6712
Epoch 1 | Train Loss: 1.1033 | Val Loss: 0.5457 | RMSE: 0.7387 | MAE: 0.6639
Epoch 2 | Train Loss: 1.0756 | Val Loss: 0.5108 | RMSE: 0.7147 | MAE: 0.6432
Epoch 3 | Train Loss: 1.0010 | Val Loss: 0.4405 | RMSE: 0.6637 | MAE: 0.5955
Epoch 4 | Train Loss: 0.8823 | Val Loss: 0.3358 | RMSE: 0.5795 | MAE: 0.5029
Epoch 5 | Train Loss: 0.7207 | Val Loss: 0.2478 | RMSE: 0.4978 | MAE: 0.4205
Epoch 6 | Train Loss: 0.5620 | Val Loss: 0.1968 | RMSE: 0.4436 | MAE: 0.3257
Epoch 7 | Train Loss: 0.4541 | Val Loss: 0.2102 | RMSE: 0.4584 | MAE: 0.3289
Epoch 8 | Train Loss: 0.4033 | Val Loss: 0.2187 | RMSE: 0.4677 | MAE: 0.3413
Epoch 9 | Train Loss: 0.3781 | Val Loss: 0.2444 | RMSE: 0.4943 | MAE: 0.3838
Epoch 10 | Train Loss: 0.3686 | Val Loss: 0.2418 | RMSE: 0.4917 | MAE: 0.3714
Epoch 11 | Train Loss: 0.3586 | Val Loss: 0.2683 | RMSE: 0.5180 | MAE: 0.4040
Epoch 12 | Train Loss: 0.3510 | Val Loss: 0.2395 | RMSE: 0.4894 | MAE: 0.3

/usr/local/lib/python3.11/dist-packages/torch_geometric/deprecation.py:26: UserWarning: 'nn.glob.GlobalAttention' is deprecated, use 'nn.aggr.AttentionalAggregation' instead
  warnings.warn(out)


Epoch 0 | Train Loss: 0.8327 | Val Loss: 1.6747 | RMSE: 1.2941 | MAE: 1.0480
Epoch 1 | Train Loss: 0.8241 | Val Loss: 1.6511 | RMSE: 1.2849 | MAE: 1.0352
Epoch 2 | Train Loss: 0.7955 | Val Loss: 1.5635 | RMSE: 1.2504 | MAE: 0.9966
Epoch 3 | Train Loss: 0.7065 | Val Loss: 1.3685 | RMSE: 1.1698 | MAE: 0.9050
Epoch 4 | Train Loss: 0.5747 | Val Loss: 1.1355 | RMSE: 1.0656 | MAE: 0.7705
Epoch 5 | Train Loss: 0.4565 | Val Loss: 0.9046 | RMSE: 0.9511 | MAE: 0.6409
Epoch 6 | Train Loss: 0.3653 | Val Loss: 0.7360 | RMSE: 0.8579 | MAE: 0.5461
Epoch 7 | Train Loss: 0.3187 | Val Loss: 0.6508 | RMSE: 0.8067 | MAE: 0.5039
Epoch 8 | Train Loss: 0.3016 | Val Loss: 0.5940 | RMSE: 0.7707 | MAE: 0.4924
Epoch 9 | Train Loss: 0.2958 | Val Loss: 0.5909 | RMSE: 0.7687 | MAE: 0.4384
Epoch 10 | Train Loss: 0.2909 | Val Loss: 0.5642 | RMSE: 0.7512 | MAE: 0.4431
Epoch 11 | Train Loss: 0.2801 | Val Loss: 0.5463 | RMSE: 0.7391 | MAE: 0.4777
Epoch 12 | Train Loss: 0.2750 | Val Loss: 0.5516 | RMSE: 0.7427 | MAE: 0.4

### 7) Evaluate

In [41]:
import pandas as pd

target = pd.read_csv("/kaggle/input/private-hen-productivity-target-labels/production_net.csv", sep=";")
eggs_per_day_df = target[["date", "Laying.rate...."]].rename(columns={"Laying.rate....": "productivity rate"})
eggs_per_day_df['productivity rate'] = eggs_per_day_df['productivity rate'].str.replace(',', '.').astype(float)
eggs_per_day_df.head(5)

date  productivity rate
0  2017-02-27          94.960274
1  2017-03-01          94.260145
2  2017-04-25          92.655367
3  2017-07-12          89.730444
4  2017-07-22          88.271474

In [42]:
mean_eggs = eggs_per_day_df["productivity rate"].mean()
std_eggs = eggs_per_day_df["productivity rate"].std()

print(f"Mean productivity rate: {mean_eggs:.2f}")
print(f"Standard deviation of productivity rate: {std_eggs:.2f}")

Mean productivity rate: 87.16
Standard deviation of productivity rate: 8.34


## Model Evaluation: Egg Productivity Prediction

### Baseline Statistics
- **Mean productivity rate:** 87.16  
- **Standard deviation:** 8.34  

---

### Model 1: Global Mean Pooling
- **Average RMSE:** 0.5359 ± 0.2512  
- **Average MAE:** 0.3239 ± 0.1025  
- Highest error among all models — suggests mean pooling may dilute critical node-level information.

---

### Model 2: Global Max Pooling
- **Average RMSE:** 0.5221 ± 0.1723  
- **Average MAE:** 0.3760 ± 0.0689  
- Slight improvement in RMSE over mean pooling — max pooling may capture dominant node signals, though MAE increased.

---

### Model 3: Global Attention Pooling
- **Average RMSE:** 0.4623 ± 0.1234  
- **Average MAE:** 0.3067 ± 0.0578  
- Best performance overall — attention mechanism helps identify and weight important nodes, improving both accuracy and consistency.

---

### Overall Insights
- Switching from **mean to attention pooling** significantly reduced prediction error.
- Attention pooling **outperformed both mean and max** in terms of RMSE and MAE.
- **Model 3’s MAE (0.3067)** is approximately **1/27th** of the standard deviation — indicating strong prediction performance relative to variability in the data.